src/deluge/io/midi/midi_follow.cpp

```diff
@@ -0,0 +1,700 @@
+    {7, 3, MIDI_CC_NONE, 10, MIDI_CC_NONE, 63, 62, MIDI_CC_NONE},
+    {5, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, 19},
+    {72, 76, 75, 73, 70, MIDI_CC_NONE, 71, 74},
+    {80, 79, 78, 77, 83, MIDI_CC_NONE, 82, 81},
+    {MIDI_CC_NONE, MIDI_CC_NONE, 61, MIDI_CC_NONE, 60, MIDI_CC_NONE, 86, 84},
+    {51, MIDI_CC_NONE, 50, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, 87, 85},
+    {58, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, 18, 17, 93, 16},
+    {59, MIDI_CC_NONE, MIDI_CC_NONE, 91, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE},
+    {53, MIDI_CC_NONE, MIDI_CC_NONE, 52, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE},
+    {MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE, MIDI_CC_NONE}};
+
+MidiFollow midiFollow{};
+
+//initialize variables
+MidiFollow::MidiFollow() {
+	init();
+}
+
+void MidiFollow::init() {
+	successfullyReadDefaultsFromFile = false;
+
+	initMapping(paramToCC);
+
+	for (int32_t i = 0; i < 128; i++) {
+		timeLastCCSent[i] = 0;
+		previousKnobPos[i] = kNoSelection;
+	}
+
+	clipForLastNoteReceived = nullptr;
+
+	timeAutomationFeedbackLastSent = 0;
+}
+
+void MidiFollow::initMapping(int32_t mapping[kDisplayWidth][kDisplayHeight]) {
+	for (int32_t xDisplay = 0; xDisplay < kDisplayWidth; xDisplay++) {
+		for (int32_t yDisplay = 0; yDisplay < kDisplayHeight; yDisplay++) {
+			mapping[xDisplay][yDisplay] = defaultParamToCCMapping[xDisplay][yDisplay];
+		}
+	}
+}
+
+/// checks to see if there is an active clip for the current context
+/// cases where there is an active clip:
+/// 1) pressing and holding a clip pad in arranger view, song view, grid view
+/// 2) pressing and holding the audition pad of a row in arranger view
+/// 3) entering a clip
+Clip* MidiFollow::getClipForMidiFollow(bool useActiveClip) {
+	Clip* clip = nullptr;
+	if (getRootUI() == &sessionView) {
+		clip = sessionView.getClipForLayout();
+	}
+	else if ((getRootUI() == &arrangerView)) {
+		if (isUIModeActive(UI_MODE_HOLDING_ARRANGEMENT_ROW) && arrangerView.lastInteractedClipInstance) {
+			clip = arrangerView.lastInteractedClipInstance->clip;
+		}
+		else if (isUIModeActive(UI_MODE_HOLDING_ARRANGEMENT_ROW_AUDITION)) {
+			Output* output = arrangerView.outputsOnScreen[arrangerView.yPressedEffective];
+			clip = currentSong->getClipWithOutput(output);
+		}
+	}
+	else {
+		clip = currentSong->currentClip;
+	}
+	//special case for instruments where you want to let notes and MPE through to the active clip
+	if (!clip && useActiveClip) {
+		clip = currentSong->currentClip;
+	}
+	return clip;
+}
+
+/// based on the current context, as determined by clip returned from the getClipForMidiFollow function
+/// obtain the modelStackWithParam for that context and return it so it can be used by midi follow
+ModelStackWithAutoParam*
+MidiFollow::getModelStackWithParam(ModelStackWithThreeMainThings* modelStackWithThreeMainThings,
```

Comment:
this would be much easier to follow if it was split into seperate functions

1. One requiring a clip and timeline counter, and one with a modelstack and no clip
2. When there is a clip, a seperate function to check params for each of audio/kit/synth/midi

Reply:
done!